assets/app/view/game/part/town_rect.rb

```diff
@@ -35,63 +35,113 @@ class TownRect < Base
 
+        MULTIPLE_STOP_TWO_EXIT_REGIONS = {
+          straight: [
+            [15],
+            [14],
+            [7],
+            [8],
+            [9],
+            [16],
+          ],
+          sharp: [
+            [15, 20, 21],
+            [12, 13, 14],
+            [0, 6, 7],
+            [2, 3, 8],
+            [9, 10, 11],
+            [16, 17, 23],
+          ],
+          gentle: [
+            [15, 12],
+            [13, 14],
+            [6, 7],
+            [2, 8],
+            [9, 10],
+            [16, 17],
+          ],
+        }.freeze
+
         EDGE_TOWN_REGIONS = {
-          straight: City::EDGE_CITY_REGIONS,
-          sharp: City::EDGE_CITY_REGIONS,
-          gentle: City::EDGE_CITY_REGIONS,
+          0 => [21],
+          0.5 => [13, 21],
+          1 => [13],
+          1.5 => [6, 13],
+          2 => [6],
+          2.5 => [2, 6],
+          3 => [2],
+          3.5 => [2, 10],
+          4 => [10],
+          4.5 => [10, 17],
+          5 => [17],
+          5.5 => [17, 21],
+        }.freeze
+
+        # bias away from top and bottom if possible
+        EDGE_TOWN_REVENUE_REGIONS = {
+          0 => [[23], false],
+          0.5 => [[12], true],
+          1 => [[5], true],
+          1.5 => [[19], false],
+          2 => [[12], false],
+          2.5 => [[5], false],
+          3 => [[0], false],
+          3.5 => [[11], true],
+          4 => [[18], true],
+          4.5 => [[23], true],
+          5 => [[11], false],
+          5.5 => [[18], false],
         }.freeze
 
         # absolute value of angle (in degrees) to tilt the town rectangle relative
         # to a straight line from edge_a to the opposite edge
         RECTANGLE_TILT = {
-          sharp: 10,
-          gentle: 10,
+          sharp: 40,
+          gentle: 15,
         }.freeze
 
         # from center of the hex, how many degrees of offset to apply to
         # positioning the town rectangle; this value will be added/subtracted to
         # (edge_a * 60)
         POSITIONAL_ANGLE = {
-          sharp: 2,
-          gentle: 2,
+          sharp: 12.12,
+          gentle: 6.11,
         }.freeze
 
         DOUBLE_DIT_REVENUE_ANGLES = [10, -130, 130, -10, 50, -50].freeze
         DOUBLE_DIT_REVENUE_REGIONS = City::OO_REVENUE_REGIONS
 
-        # Returns the two edges so that a is the edge to render close to, and
+        # Returns the two edges plus min edge so that a is the edge to render close to, and
         # (a - b).abs is 1, 2 or 3.
-        def normalized_edges
-          @normalized_edges ||=
-            begin
-              edge_a = @tile.preferred_city_town_edges[@town]
-              edge_b = (@town.exits - [edge_a]).first
-              edges = [edge_a, edge_b]
-              edges[edges.index(edges.min)] += 6 if (edge_a - edge_b).abs > 3
-              edges
-            end
-        end
-
-        def min_edge
-          @min_edge ||= normalized_edges.min
+        def self.normalized_edges(edge_a, exits)
```

Comment:
you shouldn't remove the caching here. can we make this module and have class level caching? it seems like this would be a candidate for that let me know if you need help with that

Reply:
I pulled out caching because of the re-use. Instead I just save the result where it is needed rather than calling normalized_edges over and over again.

Reply:
it will still get called for every town_rect and everything that calls these methods which will be thousands of  times

Reply:
I would need help (or at least more explanation for what you are looking for).

Reply:
sure, happy to discuss in slack